### PR TITLE
feat: introduce option to set additional getToken parameters

### PR DESCRIPTION
**Pull request by philsch**
Adds an option to configure additional parameters for `getToken` calls as described in #177

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ export interface FastifyOAuth2Options {
   credentials: Credentials;
   callbackUri: string;
   callbackUriParams?: Object;
+  getTokenParams?: Object;
```
Comment (mcollina) [inline index.d.ts:27]: can you add a test for this new type? It seems it's a function, not an object.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 1 above)
Comment (philsch) [reply index.d.ts:27]: When calling `getAccessTokenFromAuthorizationCodeFlow` those parameters are added to the payload of the `o.oauth2.getToken` POST request. I know the name `getTokenParams` is not optimal, but it is an object. 

For example if it's set to:

```
getTokenParams: {
  client_id: 'id',
  client_secret: 'secret'
}
``` 

and setting `export DEBUG=*simple-oauth2*`, then you can see that this parameters are added to the payload when retrieving the OAuth tokens:

```
 simple-oauth2:client Creating request to: (POST) /oauth/token +0ms
 simple-oauth2:client Using request options: {"headers":{"Authorization":"Basic XXX","Content-Type":"application/x-www-form-urlencoded"},"payload":"grant_type=authorization_code&client_id=id&client_secret=secret&code=3dd...71&redirect_uri=https%3A%2F%2F..."} +0ms
```

Can you think about a better naming? Maybe `tokenRequestParams?: Object`?

I'll have a look into testing in the meantime.

**Comment by philsch** [comment]
Added an additional test and renamed the parameter to `tokenRequestParams`, hope it makes more sense this way.

**Comment by philsch** [comment]
Another possible solution that came to my mind is extending `getAccessTokenFromAuthorizationCodeFlow` to support additional parameters the same like `getNewAccessTokenUsingRefreshToken(refreshToken, params, callback)` already does in a backwards compatible way, for example: 

`getAccessTokenFromAuthorizationCodeFlow(request, [params], callback)` (params being optional)

Let me know what you think about it.

File: test/plugin.test.js [hunk 2]
```diff
@@ -297,6 +297,92 @@ t.test('options.callbackUriParams', t => {
...
+    t.error(err)
+
+    fastify.inject({
+      method: 'GET',
+      url: '/callback?code=' + oAuthCode
+    }, function (err, responseStart) {
+      t.error(err)
+
+      githubScope.done()
+      t.end()
```
Comment (Eomm) [inline test/plugin.test.js:381]: Could you set the `t.plan()`?

I know the test above does not have it, but we should fix that too (it is not up to you of course)

File: index.js [hunk 3]
```diff
@@ -88,10 +98,12 @@ const oauthPlugin = fp(function (fastify, options, next) {
   }
 
   const cbk = function (o, code, callback) {
-    return callbackify(o.oauth2.getToken.bind(o.oauth2, {
+    const body = Object.assign({}, generateTokenRequestParams(tokenRequestParams), {
```
Comment (Eomm) [inline index.js:101]: I don't understand this extra call to `generateTokenRequestParams`.
It cannot be set by the user (it is hidden into `credentials.auth[kGenerateTokenRequestParams]`) and the default `defaultGenerateTokenRequestParams` returns the input object itself.

Why it is necessary?

Comment (philsch) [reply index.js:101]: I tried to implement the same principle like for `generateAuthorizationUri`. The logic there allows two ways how to set additional callback params:

* by the user via config parameter
* by the plugin maintainer for preconfigured connections (see `oauthPlugin.APPLE_CONFIGURATION`)

As said I tried to follow the same code style like for the callback parameters. If the callback implementation for `APPLE_CONFIGURATION` was a very specific workaround, I agree that this implementation is probably not needed for this PR.

Just let me know if I should simplify it.

**Review by mcollina — APPROVED** [review]
lgtm

File: index.js (hunk 3 above)
Comment (Eomm) [reply index.js:101]: Yes, please remove it as it is not necessary

Comment (philsch) [reply index.js]: done

**Review by Eomm — APPROVED** [review]

**Pull request merged**